### ISSUE TITLE
more fine-grained handling of audio corpora

### ISSUE THREAD

**Issue by genekogan** [ibab/tensorflow-wavenet#111]
this is related to [https://github.com/ibab/tensorflow-wavenet/issues/104](https://github.com/ibab/tensorflow-wavenet/issues/104).

i'm thinking ahead perhaps to when this repo supports conditioning on labels. since most interesting audio sets may not be so neatly organized as VCTK, you can potentially trim your audio and create labels via analysis. 

even without labels, it's useful to be able to train on a subset of a folder, or even subsets of individual audio files. one workflow I am developing is to isolate a segment of interest, then find audio chunks which are similar to it (using mel bands or other DSP stats useful in music information retrieval). it can be a bit heavy-handed but it can help weed out noise or other undesirable material.

it might be beyond the scope of what this repo is aiming for, in which case I'll just keep developing it separately. but one useful feature that I'd propose to start is simply letting you specify your training set in a text file rather than just a directory. maybe like a JSON file with paths to audio files and a list of subsegments to pull out. some supporting scripts can be used to generate it, e.g. a function which takes an audio file and time interval as input, and produces a JSON of T seconds of audio segments from a directory which are most similar to the input.

**Comment by ibab** [comment]
@genekogan: In the long run, I'd like this project to become more general, so that it can be used as a simple Python library in other projects.
I'm currently working on providing a `WaveNet` interface that takes care of all TensorFlow details like maintaining a session, etc. so that users can just do something like
```python
reader = MyCustomAudioReader('./audio_dir')
net = WaveNet(input=reader, **kwargs)
for i in range(1000):
    loss, _ = net.train_batch()
audio = net.sample(size=16000)
```
So you could implement a custom audio pipeline by inheriting from something like `AudioReaderBase`, and the audio reader would be able to say which extra conditioning information it has available, which would automatically get picked up by `WaveNet`.

**Comment by genekogan** [comment]
that would be super cool to see. in that case, when i have something developed, i'll share.

**Comment by lemonzi** [comment]
Should we close this an open an empty PR for the `AudioReader` refactor? And @genekogan, you can directly open a PR when you have something working.

**Comment by genekogan** [comment]
yep, agreed to that.
